tx indexing (Refs #237)

melekes opened the pull request:
* save transactions to blockstore

* move to a separate module

* benchmark KVIndexer

* batch write transactions

**Benchmarks:**

_Using golevelDB SetSync:_
```
BenchmarkKVIndexerIndex-2         100000            516300 ns/op
PASS
ok      github.com/tendermint/tendermint/blockchain/tx  56.506s

5,16 s for 10000 transactions
1 s for 2000 transactions
```

_Using golevelDB Set:_
```
BenchmarkKVIndexerIndex-2       h 3000000             8622 ns/op
PASS
ok      github.com/tendermint/tendermint/blockchain/tx  34.210s

86 ms for 10000 transactions
16 ms for 2000 transactions
```

_Using golevelDB Batch:_
```
BenchmarkKVIndexerIndex1-2               5000000              7160 ns/op
BenchmarkKVIndexerIndex500-2               20000           1750411 ns/op
BenchmarkKVIndexerIndex1000-2              10000           3573973 ns/op
BenchmarkKVIndexerIndex2000-2               5000           7836851 ns/op
BenchmarkKVIndexerIndex10000-2              1000          33438980 ns/op
PASS
ok      github.com/tendermint/tendermint/blockchain/tx  209.482s

7,8 ms for 2000 transactions
```

* [state] write test for ApplyBlock

melekes commented:
Questions:

1. Do we need to make it optional? I.e. in config `indexer = kv (default: none)`
2. What other fields besides `tx`, `height` and `DeliverTxResponse` we want to be stored?

ebuchman commented on node/node.go:
```diff
@@ -107,6 +108,12 @@ func NewNode(config cfg.Config, privValidator *types.PrivValidator, clientCreato
+	// Transaction indexing
+	store := dbm.NewDB("tx_indexer", config.GetString("db_backend"), config.GetString("db_dir"))
+	indexer := txindexer.NewKV(store)
+	bcReactor.SetTxIndexer(indexer)
```
What if we put it on `state` instead of on the two reactors? Does that make sense?

ebuchman replied:
Then we don't need to pass it to ApplyBlock since the state will already have it

melekes replied:
makes sense!

melekes replied:
But won't be there any errors in the future related to copying/loading that state?

melekes replied:
I am curious what happens when we copy something holding a reference to levelDB (with open file descriptor). Do copy shares the same file descriptor with original object?

```
func (s *State) Copy() *State {
	return &State{
		db:              s.db,
		GenesisDoc:      s.GenesisDoc,
		ChainID:         s.ChainID,
		LastBlockHeight: s.LastBlockHeight,
		LastBlockID:     s.LastBlockID,
		LastBlockTime:   s.LastBlockTime,
		Validators:      s.Validators.Copy(),
		LastValidators:  s.LastValidators.Copy(),
		AppHash:         s.AppHash,
		TxIndexer:       s.TxIndexer,
	}
}
```

melekes replied:
Ok, if we copy the pointer (not value) and do not save `TxIndexer` field, it should work fine.

ebuchman commented on state/execution_test.go:
```diff
@@ -31,6 +35,32 @@ func TestExecBlock(t *testing.T) {
 	// TODO
 }
 
+func TestApplyBlock(t *testing.T) {
```
Great we need some tests here. There was an earlier attempt to test ExecBlock (https://github.com/tendermint/tendermint/pull/328/files) but it's more complicated than it needs to be since it relies on event subscriptions - but now we can access the tx results directly!

Also, sorry, but I totally gutted this file on develop. All the handshake stuff moved to consensus/replay_test.go. So rebasing might be a bit painful around here

ebuchman commented on state/tx/indexer/kv.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+// KV is a simpliest possible indexer, backed by Key-Value storage (levelDB).
```
simplest

ebuchman commented on state/tx/indexer/kv.go:
```diff
@@ -0,0 +1,55 @@
+
+	rawBytes := indexer.store.Get([]byte(hash))
+	if rawBytes == nil {
+		return nil, nil
```
Why no ErrTxNotFound ?

melekes replied:
I think `{nil, nil}` tuple is self-describing. Plus I was trying to copy bleve interface, so later we could allow to use external indexers (https://github.com/blevesearch/bleve/blob/master/index.go#L165)

ebuchman commented on state/tx/indexer/kv_test.go:
```diff
@@ -0,0 +1,62 @@
+	indexer := &KV{store: db.NewMemDB()}
+
+	tx := types.Tx("HELLO WORLD")
+	txResult := &types.TxResult{tx, 1, abci.ResponseDeliverTx{Data: []byte{0}, Code: abci.CodeType_OK, Log: ""}}
```
`abci.NewResultOK(data []byte, log string)`

melekes replied:
I tried `abci.NewResultOK([]byte{0}, "").(abci.ResponseDeliverTx)`, but got an error because `abci.Result` is a non-interface type.

ebuchman commented on types/tx_result.go:
```diff
@@ -0,0 +1,15 @@
+// TxResult contains DeliverTx response and height of the block, where this
+// transaction was committed.
+//
+// One usage is indexing transaction results.
```
Should probably add `Index` for the txs position in the block at Height

melekes commented:
Thanks for the review!

Rebased. It was easier than I thought it will be :) Just needed to remember `git rebase develop -Xtheirs`

Also removed `Tx` field from TxResult.

ebuchman commented:
Nice trick! Can we get a config option on this, off by default for now?

melekes commented:
Sure! Also, I was thinking.. do we need to include a `BlockHash` (similar to Ethereum)

ebuchman commented:
In the `TxResult`? I don't think so because we have the height and we don't fork (blocks are unique per height). That said, for sake of recovery from a double-spend attack where +1/3 are malicious, we may need that. But I think that's something we'll deal with separately - so maybe just make a note of it

melekes commented:
Will do config option, RPC, and a new header field tomorrow.

melekes commented:
```
EROR[02-22|12:12:34] Panic in RPC HTTP handler                module=rpcserver error="Paniced on a Sanity Check: Unexpected nil-pointer of type types.TxResult for registered interface TMResult. For compatibility with other languages, nil-pointer interface values are forbidden." stack="goroutine 72 [running]:\nruntime/debug.Stack(0xa242a0, 0x99d240, 0xc421dfbe40)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x79\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server.RecoverAndLogHandler.func1.1(0xc421e01120, 0xed03f7732, 0xd3ca36a, 0xec28c0, 0xc4211f5680)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server/http_server.go:90 +0x70b\npanic(0x99d240, 0xc421dfbe40)\n\t/usr/local/go/src/runtime/panic.go:458 +0x243\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-common.PanicSanity(0x9c1880, 0xc421dfbe20)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common/errors.go:26 +0xe0\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire.JSONBytes(0x97ec40, 0xc421dfbe00, 0x16, 0x94, 0x94)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-wire/util.go:33 +0x116\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/types.NewRPCResponse(0x0, 0x0, 0x97ec40, 0xc421dfbe00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/types/types.go:55 +0x141\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server.makeHTTPHandler.func2(0xe927a0, 0xc421e01120, 0xc4211f5680)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server/handlers.go:218 +0x697\nnet/http.HandlerFunc.ServeHTTP(0xc42049bbe0, 0xe927a0, 0xc421e01120, 0xc4211f5680)\n\t/usr/local/go/src/net/http/server.go:1726 +0x44\nnet/http.(*ServeMux).ServeHTTP(0xc420874000, 0xe927a0, 0xc421e01120, 0xc4211f5680)\n\t/usr/local/go/src/net/http/server.go:2022 +0x7f\ngithub.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server.RecoverAndLogHandler.func1(0xe92b20, 0xc421dea5b0, 0xc4211f5680)\n\t/home/vagrant/go/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-rpc/server/http_server.go:108 +0x35a\nnet/http.HandlerFunc.ServeHTTP(0xc4208556a0, 0xe92b20, 0xc421dea5b0, 0xc4211f5680)\n\t/usr/local/go/src/net/http/server.go:1726 +0x44\nnet/http.serverHandler.ServeHTTP(0xc42009a300, 0xe92b20, 0xc421dea5b0, 0xc4211f5680)\n\t/usr/local/go/src/net/http/server.go:2202 +0x7d\nnet/http.(*conn).serve(0xc421d24c80, 0xe935e0, 0xc421d2cb40)\n\t/usr/local/go/src/net/http/server.go:1579 +0x4b7\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:2293 +0x44d\n"
```

Internal Server Error: Paniced on a Sanity Check: Unexpected nil-pointer of type types.TxResult for registered interface TMResult. For compatibility with other languages, nil-pointer interface values are forbidden.

Keep getting this error...

melekes commented:
The only thing left here - "getting the tx results hash into the header" of the next block presumably. Also, writing integration tests for RPC call would not hurt.

melekes commented:
Been looking at the code. So, in order for us to include tx results hash into the header we need:
1) add `TxResultsHash` to state/state.go
2) after we successfully applied block, set `TxResultsHash` (like we do with AppHash)
3) during construction of the next block, include `TxResultsHash` into the block header

Q: should it be a separate PR?

melekes commented:
RPC examples:

```
# with indexing enabled
$ curl -s 'localhost:46657/tx?hash="DD2E325F797788759D278C1D4B370C842E"'
{"jsonrpc":"2.0","id":"","result":[99,{"height":2454,"index":0,"deliver_tx":{}}],"error":""}

# with indexing disabled (default now)
$ curl -s 'localhost:46657/tx?hash="DD2E325F79F7E5F77788759D278C1D4B370C842E"'
{"jsonrpc":"2.0","id":"","result":null,"error":"Error unreflecting result: Indexing is disabled (set `tx_indexer=kv` in config)"}
```

ebuchman commented:
Thanks Anton. The aim is to get this into 0.9.0 or 0.9.1. Since the TxResultsHash in the header will be breaking, it will have to wait until 0.10.0. So another PR is fine. We have some other changes to the header to make for 0.10.0 as well (like NextValidatorsHash), so there's no rush yet.

melekes commented:
**With tx indexing**
```
$ tm-bench -T 60 -r 1000 -c 10 127.0.0.1:46657
Stats             Avg        Stdev      Max
Block latency     0.21ms     0.01ms     0ms
Blocks/sec        0.678      0.467      1
Txs/sec           6450       4776       10000
```

**Without tx indexing**
```
$ tm-bench -T 60 -r 1000 -c 10 127.0.0.1:46657
Stats             Avg        Stdev      Max
Block latency     0.22ms     0.01ms     0ms
Blocks/sec        0.678      0.467      1
Txs/sec           6610       4734       10000
```

ebuchman commented on state/execution.go:
```diff
@@ -73,21 +80,28 @@ func execBlockOnProxyApp(eventCache types.Fireable, proxyAppConn proxy.AppConnCo
+
+			tx := types.Tx(req.GetDeliverTx().Tx)
+			index, ok := txHashToIndexMap[string(tx.Hash())]
+			if ok {
```
how could we have not ok? is this map really necessary? these callbacks should be called in order so we should just be able to use a simple counter. Or am I missing something?

ebuchman commented:
Is there a reason we index using the hex-encoding rather than the underlying bytes?

melekes commented:
I was looking for inspiration in a number of clients (elastic, ...), engines like bleve and most of them use string keys (https://github.com/blevesearch/bleve/blob/master/index.go#L154)

I guess we can switch to byte keys to avoid unnecessary transformations.

ebuchman commented:
Ah ok. Would we be compromising some kind of interoperability by doing that? I would expect most of these other indexers using strings because thats the natural form of their indices. Whereas the natural form for us really is []byte.

melekes commented:
It depends on how are we gonna support external indexers. If there will be an adapter/client per index engine:

```
import "github.com/olivere/elastic"

func (indexer *indexers.Elastic) Index(key []byte, value..) {
  txs.Index().Index(string(key)).
}
```

then having a byte array key is ok.

If it will be in some form of plugins, where we have a common interface, then I guess the string key makes more sense.

ebuchman commented:
I imagine the former is more flexible. Cool

ebuchman commented on types/tx_result.go:
```diff
@@ -0,0 +1,14 @@
+//
+// One usage is indexing transaction results.
+type TxResult struct {
+	Height    uint64                 `json:"height"`
```
is there a reason these aren't just `int` ?

melekes replied:
I think we've already discussed this before. `uint` because it is never negative and only goes up. `64` because we want to preserve the size (ie. 1st node has 386 arch, 2nd node amd64)

ebuchman replied:
right. I feel like we did discuss before but I didn't remember conclusion. We use `int` all over the codebase where we could/should use `uint64` so I wonder if we should break with that or at some point just fix it everywhere. So what about the `uint32` for the index? 

@jaekwon

melekes replied:
We decided to use `int` in places where it can go down (because otherwise it will require checks for underflow). But here `Height` only goes up, so it is OK.

`32` 4294967295 vs `64` 18446744073709551615 :)

ethanfrey replied:
Yes, please formalize this somewhere and adjust things.

I have way to many `h := height.(uint64)` and `height := h.(int)` lines... Since abci app returns `int64` for block height and tendermint rpc returns `int`.  So when I want to compare a proof with a header....

I don't care much about the choice (though I like anton's reasoning).  But I would really appreciate consistency, even if that doesn't make it in until 0.10, at least let's standardize it all on the `unstable` branch.

ebuchman commented on state/execution.go:
```diff
@@ -218,26 +224,32 @@ func (s *State) validateBlock(block *types.Block) error {
+
+	batch := txindexer.NewBatch()
+	for i, r := range txResults {
+		if r != nil {
```
is this ever nil, now that we assume the ABCI return vals are in order and complete?

melekes replied:
we can remove it, yes

ebuchman commented on rpc/core/types/responses.go:
```diff
@@ -67,11 +67,18 @@ type ResultBroadcastTx struct {
+}
+
+type ResultTx struct {
+	types.TxResult
```
Include the Tx itself in here

ebuchman commented:
Ok. This is looking good. I'll address the above comments later tonight or tomorrow morning. Just want to hear from @melekes about the `int` s. 

This will close https://github.com/tendermint/tendermint/issues/237

It also sort of addresses https://github.com/tendermint/tendermint/issues/415 . Should we still add the `Height` to `ResultBroadcastTxCommit` ? It wouldn't hurt but we can just fetch it now with the `/tx` call. 

Also, not sure we need https://github.com/tendermint/tendermint/issues/423 now if we can just use `/tx`. 

As for https://github.com/tendermint/tendermint/pull/441, maybe we should add a `proof bool` to the `/tx` call to request a proof the tx was included in a block?

melekes commented:
> Also, not sure we need #423 now if we can just use /tx.

if we include the tx itself, then we won't be needing a separate RPC call.

> As for #441, maybe we should add a proof bool to the /tx call to request a proof the tx was included in a block?

yes

melekes commented:
> It also sort of addresses #415 . Should we still add the Height to ResultBroadcastTxCommit ? It wouldn't hurt but we can just fetch it now with the /tx call.

@ethanfrey what was the use case behind #415? when you get a height, are you going to fetch the whole block?

ethanfrey commented:
The use case behind #415 is that I can query the block chain for a complete proof of the TX. Also, if I want to query for the resulting state, I want to know TX is in block h, so I can query/wait for block h+1

It would also be good in the TX event, so if I use async, the notification message tells me what block it was in. 

I want to remove any potential race conditions there. In a clean way. This comes up in test code a fair bit. When we actually have smart clients, it will arise there as well. 

Allowing me to look up the TX by hash to find the block height is also very useful. More more for history. Like checking up on txs from several minutes or hours ago. The height in the response is very useful for responsive clients.

melekes commented:
> The height in the response is very useful for responsive clients.

ok, let's do both https://github.com/tendermint/tendermint/issues/415#issue-209219532 and https://github.com/tendermint/tendermint/issues/415#issuecomment-281471628

ebuchman commented:
Ok I added the Height stuff. The problem with https://github.com/tendermint/tendermint/issues/415#issuecomment-281471628 is it would be breaking (on the event structure) so we have to wait for 0.10. Also, @ethanfrey do we need the tx index in ResultBroadcastTxCommit, or just the height is fine?

ebuchman commented:
I merged https://github.com/tendermint/tendermint/pull/441 in and added the `prove` flag to `/tx`. I believe we're done here. 

Could use a review from @jaekwon before merge

melekes commented on rpc/core/tx.go:
```diff
@@ -0,0 +1,35 @@
+		return &ctypes.ResultTx{}, fmt.Errorf("Tx (%X) not found", hash)
+	}
+
+	block := blockStore.LoadBlock(int(r.Height))
```
Just to make sure, we could have Tx as a field of `TxResult` 

```
txResults[txIndex] = &types.TxResult{uint64(block.Height), uint32(txIndex), **txBytes**,*txResult}
```

but this would mean we're duplicating tx bytes (storing them in 2 places) and we do not want that, correct?

ebuchman replied:
Yes I think we want to avoid that.

ethanfrey commented:
@ebuchman just the height is fine. If anyone wants to look up the whole data later they can do that with /tx Since #412 is closed, I don't see such use for the TX index.

ethanfrey commented:
But if I am missing something, feel free to add it

ethanfrey commented:
With the proof flag when we query by hash, the index is totally unneeded for any use case I have. This is awesome!!! 

Thank you bucky and Anton. This will remove my biggest complaints with the api.

ethanfrey commented:
Okay, trying to use this code and I found two little details, maybe they can be cleaned up before releasing it.

First, I'd love consistency between int/uint64 for height.  At least within one file:
https://github.com/tendermint/tendermint/blob/feature/237-tx-indexing/rpc/core/types/responses.go#L78-L82

Second, if I want to have any client code use this feature, I need a machine-parseable way to determine if the feature is enabled on the server.  I think parsing the error string to look for the words "Indexing is disabled" in the error message is sub-optimal.  Maybe add a boolean flag to `/status` or somewhere else? I don't really care where, just something machine readable.

ebuchman commented:
I like the idea of adding a list of enabled features to the `/status`. will do

jaekwon commented on rpc/core/tx.go:
```diff
@@ -0,0 +1,81 @@
+		return nil, fmt.Errorf("Invalid height (%d) for blockStore at height %d", height, blockStore.Height())
+	}
+
+	block := blockStore.LoadBlock(height)
```
Do we need to load a block if we have the txid?

I feel that we're overloading this function... I'd rather we only allowed calling this function if tx indexing is enabled.

ebuchman replied:
> Do we need to load a block if we have the txid?

Yes because txs aren't stored with the result since then they'd be stored twice. So if you want the actual tx, you need to load a block.

> I feel that we're overloading this function... I'd rather we only allowed calling this function if tx indexing is enabled.

I agree - as I was saying on call today. The idea here was that if we don't have the indexer but we do have the txid, we can still search the block for it. That still seems like useful functionality. Maybe we should separate into two calls tho: `/tx` and `/tx_result`. The former will always work - if the indexer is active, it can fetch the height/index and get the tx from the block, else it requires a height be sent as an arg and will search the block for it. The latter will only work if the indexer is active and will only return the result, no tx. Or we could add a boolean `tx` arg to `tx_result` to specify if you want the `tx` bytes too

jaekwon commented on state/execution_test.go:
```diff
@@ -0,0 +1,90 @@
+//----------------------------------------------------------------------------
+
+// make some bogus txs
+func txsFunc(blockNum int) (txs []types.Tx) {
```
Weird name

jaekwon replied:
s/txsFunc/makeTxs/g

jaekwon commented on state/tx/indexer/kv.go:
```diff
@@ -0,0 +1,55 @@
+
+// KV is a simplest possible indexer, backed by Key-Value storage (levelDB).
+// It could only index transaction by its identifier.
+type KV struct {
```
This naming doesn't seem right.  Of course Golang encourages you to have the module name as part of an exported thing's name, e.g. json.Encoder rather than json.JSONEncoder, but in this case, the struct isn't a KV, it's an Indexer.  So unless we're going to have a separate package named 'tendermint/indexer/kv', we should call this KVIndexer.

ebuchman replied:
Why don't we name it after what's being used. In this case, `indexer.GoDB`. then later `indexer.ElasticSearch`, etc.

jaekwon replied:
They're all misnomers.

ebuchman replied:
So really the correct thing would be a package for each implementation with a struct called `Indexer` ?

jaekwon commented on state/tx/indexer/kv.go:
```diff
@@ -0,0 +1,55 @@
+
+// Tx gets transaction from the KV storage and returns it or nil if the
+// transaction is not found.
+func (indexer *KV) Tx(hash []byte) (*types.TxResult, error) {
```
'GetTx'?  Or just 'Get'.  Just 'Tx' seems weird.

melekes replied:
`Get` is a good idea, yes

jaekwon commented on state/tx/indexer.go:
```diff
@@ -0,0 +1,21 @@
+
+	// Tx returns specified transaction or nil if the transaction is not indexed
+	// or stored.
+	Tx(hash []byte) (*types.TxResult, error)
```
Rename to 'GetTx'

melekes replied:
just `Get`?

jaekwon replied:
Sure

jaekwon commented on state/tx/indexer.go:
```diff
@@ -0,0 +1,21 @@
+	// NOTE We do not specify Index method for analyzing a single transaction
+	// here because it bears heavy perfomance loses. Almost all advanced indexers
+	// support batching.
+	Batch(b *txindexer.Batch) error
```
Rename to 'WriteBatch'

melekes replied:
or maybe `IndexBatch`?

jaekwon replied:
indexer.IndexBatch seems redundant.... AddBatch, WriteBatch, SetBatch, CommitBatch etc all seem ok

jaekwon commented on node/node.go:
```diff
@@ -360,6 +376,7 @@ func makeNodeInfo(config cfg.Config, sw *p2p.Switch, privKey crypto.PrivKeyEd255
 			cmn.Fmt("p2p_version=%v", p2p.Version),
 			cmn.Fmt("consensus_version=%v", consensus.Version),
 			cmn.Fmt("rpc_version=%v/%v", rpc.Version, rpccore.Version),
+			cmn.Fmt("tx_indexer=%v", config.GetString("tx_indexer")),
```
I don't think we should expose the method of indexing.  Just 'tx_index=on' would be sufficient.

melekes replied:
But what if later we will want to support external indexers (e.g. https://github.com/blevesearch/bleve)?

jaekwon replied:
i don't think we need to expose that to peers.  to my peer nodes, all they need to know is that i support some index.

ebuchman replied:
Oh - the intention here was actually more about a local client than for a peer. But it shouldn't matter much to them either I guess

jaekwon replied:
I mean, peers have no use for those field values.  What should we do when someone implements a new type of index, like "redis"?  The behavior is undefined.

If you want, we can create two fields, like "tx_index=on", "tx_index_backend="kv" or something, for diagnostic info.  I imagine that it's just exposing vulnerabilities though.

jaekwon commented on rpc/client/httpclient.go:
```diff
@@ -160,6 +160,21 @@ func (c *HTTP) Commit(height int) (*ctypes.ResultCommit, error) {
 	return (*tmResult).(*ctypes.ResultCommit), nil
 }
 
+func (c *HTTP) Tx(hash []byte, height, index int, prove bool) (*ctypes.ResultTx, error) {
```
s/Tx/GetTx/g

jaekwon commented on rpc/client/httpclient.go:
```diff
@@ -160,6 +160,21 @@ func (c *HTTP) Commit(height int) (*ctypes.ResultCommit, error) {
+		"hash":   hash,
+		"prove":  prove,
+	}
+	_, err := c.rpc.Call("tx", query, tmResult)
```
s/"tx"/"get_tx"/g

jaekwon commented on rpc/core/routes.go:
```diff
@@ -19,6 +19,7 @@ var Routes = map[string]*rpc.RPCFunc{
 	"genesis":              rpc.NewRPCFunc(GenesisResult, ""),
 	"block":                rpc.NewRPCFunc(BlockResult, "height"),
 	"commit":               rpc.NewRPCFunc(CommitResult, "height"),
+	"tx":                   rpc.NewRPCFunc(TxResult, "hash,height,index,prove"),
```
s/"tx"/"get_tx"/g

jaekwon commented on rpc/core/types/responses.go:
```diff
@@ -38,6 +40,19 @@ type ResultStatus struct {
+	}
+	for _, s := range s.NodeInfo.Other {
+		info := strings.Split(s, "=")
+		if len(info) == 2 && info[0] == "tx_indexer" {
```
I'd change this to just check for "tx_index=on", rather than 'kv' or whatnot.

jaekwon commented on rpc/core/types/responses.go:
```diff
@@ -67,11 +82,23 @@ type ResultBroadcastTx struct {
+	Height    int                     `json:"height"`
+}
+
+type ResultTx struct {
```
s/ResultTx/ResultGetTx/g

jaekwon commented on node/node.go:
```diff
@@ -84,6 +87,17 @@ func NewNode(config cfg.Config, privValidator *types.PrivValidator, clientCreato
 	state = sm.LoadState(stateDB)
 
+	// Transaction indexing
+	var txIndexer tx.Indexer
```
var txi?

jaekwon commented on rpc/core/tx.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+func Tx(hash []byte, height, index int, prove bool) (*ctypes.ResultTx, error) {
```
s/Tx/GetTx/g

jaekwon commented on state/tx/indexer/batch.go:
```diff
@@ -0,0 +1,32 @@
+}
+
+// Index adds or updates entry for the given hash.
+func (b *Batch) Index(hash []byte, result types.TxResult) error {
```
s/Index/Set/g

melekes replied:
But we're not setting anything... Maybe `Add` if you don't like the word index. Also, `Tx` seems redundant since we already have the subject in the package name (`txindexer`). Don't you think?

jaekwon replied:
Sure, works for now.

jaekwon commented on state/tx/indexer/batch.go:
```diff
@@ -0,0 +1,32 @@
+// perform operations on a batch from a single thread at a time. Once batch
+// execution has started, you may not modify it.
+type Batch struct {
+	Ops map[string]types.TxResult
```
I think we should just replace this with a slice... it'll be faster and more deterministic.  We shouldn't get duplicate txs in the same batch anyways.

jaekwon commented on state/tx/indexer/kv.go:
```diff
@@ -0,0 +1,55 @@
+package indexer
```
I'd change the path into state/txindexer/kv/kv.go . 
Also, I'd rename state/tx into state/txindexer .
This would address my other comment about renaming the KV struct into KVIndexer.
